Query refactoring: QueryFilterBuilder and Parser

cbuescher opened the pull request:
Moving the query building functionality from the parser to the builders
new toQuery() method analogous to other recent query refactorings. In
this case this also includes FQueryFilterParser, since both queries are
closely related.

Relates to #10217
Closes #11729

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryFilterBuilder.java:
```diff
@@ -78,6 +97,58 @@ protected void buildFQuery(XContentBuilder builder, Params params) throws IOExce
+        if (queryName == null) {
+            return new QueryFilterBuilder(innerQueryBuilder);
+        } else {
+            return new FQueryFilterBuilder(innerQueryBuilder).queryName(queryName);
```
I wish we could avoid having this if, it's a PITA really, I would consider breaking java api bw comp here and move the queryName to FQueryFilterBuilder so the code is more human. Thoughts?

cbuescher replied:
What if we accept a little bit of code duplication and separate FQueryFilterBuilder and QueryFilterBuilder in so far as they both extend QueryBuilder. Then FQueryBuilder would be the one that has a `queryName`. The only thing that remains slightly odd then is that when rendering out, FQuery still uses the QueryFilterBuilder name ("query") because it essentially wraps it. But thats the state of the DSL at the moment.

javanna commented:
left one comment, looks good!

cbuescher commented:
With last commit I moved the queryName member which was just used for FQuery from the QueryFilterBuilder to the new FQueryFilterBuilder that we introduced on the feature branch to separate the two variants of this builder. This creates some code duplication but avoids the former confusion between the two cases. Only problem here might be that for the Java-API we slightly break the existing QueryFilterBuilder because now it doesn't support queryName() any longer, but that part conceptually belonged to FQuery anyway.

javanna commented on core/src/test/java/org/elasticsearch/index/query/FQueryFilterBuilderTest.java:
```diff
@@ -0,0 +1,83 @@
+    /**
+     * test wrapping an inner filter that returns null also returns <tt>null</null> to pass on upwards
+     */
+    public void testInnerQueryReturnsNull() throws IOException {
```
missing @Test

s1monw replied:
we should `@Test` to forbidden API

javanna commented on core/src/test/java/org/elasticsearch/index/query/QueryFilterBuilderTest.java:
```diff
@@ -0,0 +1,74 @@
+    /**
+     * test wrapping an inner filter that returns null also returns <tt>null</null> to pass on upwards
+     */
+    public void testInnerQueryReturnsNull() throws IOException {
```
missing @Test here too

s1monw replied:
bullshit :hammer:

javanna replied:
as I mentioned a million times now, when we will say let's not use it and remove it everywhere I will stop whining about it. I don't want to see three methods with annotations and one without it, it hurts my eyes. Either with or without it, not in the middle, at least in the same test class.

javanna commented:
LGTM besides the few comments I left. Looking forward to seeing these two filters removed to be honest, with the merge of filters and queries they make very little sense (only bw comp). I think we need to create a migrate_query_refactoring.asciidoc filter and add our first breaking change (only for the java api).

cbuescher commented:
@javanna added asciidoc to track breaking changes on the feature branch, mind to have a look if you think this is a good location and this is verbose enough to keep track? Also, this is inteded for internal use mostly, should be merged with other migration doc when we merge with master, right?

javanna commented:
looks great. Yes it will eventually be merged but given that we have one file per version and we don't know yet when we will merge the branch, let's just keep a separate file and decide later on what to do with it.

javanna commented:
This change is breaking for the java api as we removed the setter `queryName(String queryName)` since this field is not supported in this type of query. Use `FQueryFilterBuilder.queryName(String queryName)` instead when in need to wrap a named query as a filter.
